### PR TITLE
[MAINTENANCE] Incorporate ValidatorGraph Populating/Resolution Methods in ValidationGraph Module

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Consolidate all `ValidationGraph` methods (e.g., `build_metric_dependency_graph()`, `resolve_validation_graph()`, etc.) and make appropriate methods in that class, such as `_parse()`, as private methods (no publicly-valuable usefulness).
* Update all references the `ValidationGraph` methods to point to this object (which now contains not only the state of the graph, but encapsulates all the methods, pertinent to this entity).
* Update tests to exercise the more effective object references properly.
* Move tests to modules that is most appropriate for their purpose.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1103/GREAT-1318
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Review by donaldheppner — CHANGES_REQUESTED** [review]
We're not ready for this change; the progress bar needs to be factored out, and computing metrics are not the concern of the `ValidationGraph` -- the graph should still be about graphing only.

Do not merge.

**Comment by alexsherstinsky** [comment]
> 

@donaldheppner One comment: There is no computing metrics in the `ValidationGraph` right now.  There is, however, a call to `resolve_validation_graph()`, because it is an operation on the graph (that is why that method has the `self` reference -- so that it can parse the graph).  It, in turn, calls `execution_engine.resolve_metrics()` -- which is how the metrics get computed.  So this is both, graph operations, and graph resolution computation.  If you have ideas for a better module for these operations, I am happy to discuss.  Thanks.

File: tests/validator/test_validation_graph.py
Comment (Kilo59) [inline tests/validator/test_validation_graph.py]:
```suggestion
    class DummyPandasExecutionEngine:
```

I would call it a dummy, not a fake.
"Fakes" are typically full implementations.
https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs

>Dummy objects are passed around but never actually used. Usually they are just used to fill parameter lists.

>Fake objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an [in memory database](https://martinfowler.com/bliki/InMemoryTestDatabase.html) is a good example).

Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @Kilo59 I called this one "Fake" because it does a little bit more than a Dummy in the sense that it sets the type of the object.  Do you still think that it should be "Dummy"?  Thanks!

Comment (Kilo59) [reply tests/validator/test_validation_graph.py]: I see, on second look this, might be more of a "Stub"

>Stubs provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.

`PandasExecutionEngineStub`

Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @Kilo59 Thank you very much -- updated!

File: tests/validator/test_validation_graph.py
Comment (billdirks) [inline tests/validator/test_validation_graph.py]: Is this an integration test or a unit test?

File: tests/validator/test_validation_graph.py
Comment (billdirks) [inline tests/validator/test_validation_graph.py]: Why rename?

File: tests/validator/test_validation_graph.py
Comment (billdirks) [inline tests/validator/test_validation_graph.py]: Does it matter that it's a `Pandas` here? If so, can you just use the real execution engine? That is 
```
import ge

ge.execution_engine.PandasExecutionEngine() 
```
instantiates a real execution engine. What does the stub get you here?

File: tests/validator/test_validation_graph.py
Comment (billdirks) [inline tests/validator/test_validation_graph.py]: This seems like the part you want to test based on the test name. The part above seems like setup that is repeated in a few tests. Could you move that out into its own helper setup function that each of the tests call?

**Review by billdirks — COMMENTED** [review]
I only looked at one test and left a few comments.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @billdirks Great catch -- all fixed.  Thanks!

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @billdirks We need the engine to be Pandas for the logic of the test to work properly -- so this stubbing achieves the purposes without instantiating the actual `PandasExecutionEngine` class.  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @billdirks This is in the name of using as little resource as possible to focus on the matter of the test; so we do this -- the execution engine we need is much less than the full `PandasExecutionEngine` instance, but a little bit more than a "Dummy" one.  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @billdirks Updated -- great idea -- thanks!

File: great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py
Comment (donaldheppner) [inline great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py:44]: Why is this here?

File: great_expectations/validator/validation_graph.py
Comment (donaldheppner) [inline great_expectations/validator/validation_graph.py:15]: why is this here?

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: this parameter is unused

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: this code and class is duplicated in a number of tests, consider making it a fixture

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: I don't think this test is necessary; there is no complexity in the constructor and the code is covered in every other test

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: this declaration can be combined with the assignment

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: seems strange to assign graph of an Error, consider another variable here for readability

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: I would not expect to be referencing the Validator in a test of the Validation Graph

Comment (donaldheppner) [reply tests/validator/test_validation_graph.py]: this variable is unused; will this test pass without it? If not, why not (that would be kind of scary)?

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: we can include a type-hint here of `dict[str, MetricConfiguration]`

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: is this meant to be the aborted metric? consider naming `bad_configuration`

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py:353]: These Dummys were called Stubs in the previous tests; I think Dummy is the more accurate term for both places

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py:363]: this series of mocks is duplicated between this method and the next; consider factoring into its own method or fixture

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py:229]: This is a dummy

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py:100]: I find this parameterization confusing. I would rather see this broken into two tests, one for table, one for column, and have a (non-test) method contain the assertions to reduce the duplication.

On deeper inspection, this is testing a simple assignment constructor. This test, `test_MetricEdge_init`, can be removed

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner @Kilo59 specifically did not want a fixture, until we all had a chance to discuss.  Can we put it up on the agenda soon?  Thank you.

File: great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py
Comment (alexsherstinsky) [reply great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py:44]: @donaldheppner Because we are now using  `graph.resolve_validation_graph()` correctly and consistently, which is defined as is currently in this module -- and used with the logging.  This correct usage is in "great_expectations/validator/metrics_calculator.py" starting on line 178.  Thanks.

File: great_expectations/validator/validation_graph.py
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py:15]: @donaldheppner Same reason as above.  We are now following the consistent usage pattern, as defined in "great_expectations/validator/metrics_calculator.py" starting on line 178. Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner Thanks for the good catch -- I deleted it (in next commit).

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner No, it was a typo -- thanks for alluding me to it -- fixed (in next commit).

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner Fixed!

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner This is used for the side-effect of using the `Validator` instantiation -- it gets `BatchData` loaded.  I believe I can come up with a more straightforward way of doing this (given our recent results with refactoring `BatchManagement` concerns out of `Validator` and others).  I will do this first thing on Monday morning (10/24/2022).  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner Added proper type hints (in next commit).  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner This is meant to indicate an `ExpectationConfiguration`, which will depend on a metric that will fail, because it will be asked to access a column that does not exist.  Do we still want to change anything in the variable names?  I feel it is not necessary, but happy to discuss.  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:353]: @donaldheppner They are different.  Dummies here are true Dummies -- they are just placeholders.  However, the Stubs here actually do something (one thing, to be precise, they ensure that the object's type will be set as needed).  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:363]: @donaldheppner @Kilo59 was explicitly against doing this; we should discuss together.  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:229]: @donaldheppner According to our logic, it is a stub.  Thanks. /cc @Kilo59

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner I agree -- @cdkini do you mind if I delete this test, or if you created with a specific purpose in mind, please let us know soon?  Thanks!

File: tests/validator/test_validation_graph.py
Comment (Kilo59) [reply tests/validator/test_validation_graph.py]: My request not to have it in a fixture was because I think the value of being to clearly see that the Dummy doesn't define any behavior (and is a true "dummy") outweighs the minor duplication. Although I'm not sure if this example is really a Dummy or Stub.

In general do think we should use fixtures and remove duplication when the Double (Mock, Stub, Fake etc.) becomes more complex. Although I don't know where the line is there. I think we should decide this as a team.

File: tests/validator/test_validation_graph.py
Comment (Kilo59) [reply tests/validator/test_validation_graph.py]: @alexsherstinsky can you explain what the stub actually does?
Why does it need to have it's `__name__` modified like that? Does the test break if you leave it un-modified?

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner Deleted (in next commit) after confirming with @cdkini.  Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:100]: @donaldheppner I agree -- @cdkini do you mind if I delete this test, or if you created with a specific purpose in mind, please let us know soon? Thanks!

Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:100]: @donaldheppner Deleted (in next commit) after confirming with @cdkini. Thanks.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @Kilo59 The goal in this and similar situations in this module, where I declared `PandasExecutionEngineStub`, is for the line 2214 in `get_validation_dependencies()` in `great_expectations/expectations/expectation.py` to return `True` -- we need the smallest possible object that can be perceived as having the type of `PandasExecutionEngine` without needing any other behaviors from it.  Can you recommend a better way, or is this a good way to do this?  Thanks!

File: great_expectations/validator/validation_graph.py
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py]: ## Nit

Why do we use `@propety` here?
Functionally this is the same as just having a `self.execution_engine` attribute that we access and modify directly.

We aren't actually protecting the underlying `self._execution_engine`. So it seems like it's just more lines of code for no obvious benefit.

If it's about preserving typing details we can just do this...

```python
class ValidationGraph:
   execution_engine: ExecutionEngine

    def __init__(self, ..., execution_engine: ExecutionEngine) -> None:
        ...
        self.execution_engine = execution_engine
        
    ...

```

As an aside, the typing appears wrong. If `execution_engine` could be `None` as part of `__init__`, then the properties can also be `None`.

Can we type_check this module?

File: great_expectations/validator/validation_graph.py
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py:46]: Can `execution_engine` really be `None`? Should this be required?

File: tests/validator/test_validation_graph.py
Comment (Kilo59) [reply tests/validator/test_validation_graph.py]: No, that seems reasonable. Thanks for the explanation.

The only suggestion I would have is to define it like this, but this might not work.
https://stackoverflow.com/a/11769062/6304433

```python
class PandasExecutionEngineStub:
   __name__ = "PandasExecutionEngine"
```

Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @Kilo59 I tried this before -- and tried it again just now -- it does not work, because the underlying code I mentioned recognizes the class as of the `PandasExecutionEngineStub` type, instead of as `PandasExecutionEngine` type.

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner I was able to remove the need to instantiate the `Validator` -- instantiating the `PandasExecutionEngine` with appropriate arguments is necessary and sufficient.  Thanks.

File: great_expectations/validator/validation_graph.py
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py]: @Kilo59 Any idea why this module is not getting type checked automatically?  What should be added and where to force type checking on new modules that we create?  I am looking into your other questions in this comment.  Thank you!

Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py]: @Kilo59 This property is not used, so I deleted the related code.  Thanks for the great catch!

File: great_expectations/validator/validation_graph.py
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py:46]: @Kilo59 Yes, the `execution_engine` here is optional (as in "not always needed").  Thanks.

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py]: Consider using pytest.raises instead: https://docs.pytest.org/en/stable/reference/reference.html?highlight=raises#pytest.raises

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py]: @donaldheppner Good point (incorporated/committed).  Thanks!

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [inline tests/validator/test_validation_graph.py:252]: I think the code paths here can be tested without an integration test; we're not trying to test that the ExecutionEngine raises errors, we are trying to test that the ValidationGraph properly handles those errors.

This would be a good case for a Mock

File: tests/validator/test_validation_graph.py
Comment (donaldheppner) [reply tests/validator/test_validation_graph.py:363]: I'm fine with it not being a fixture, but a well named method here is then the way to go. Duplication in tests is still duplication, and there is value in knowing that both tests use the same setup

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: tests/validator/test_validation_graph.py
Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:252]: @donaldheppner Investigating....

Comment (alexsherstinsky) [reply tests/validator/test_validation_graph.py:252]: @donaldheppner This has been done.  Thanks.

File: great_expectations/validator/validation_graph.py
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py:1]: @alexsherstinsky before merging this can you enable type-checking for this module?
Previously, this module only had 2 out-standing typing issues.

https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L146
To enable type-checking you need to remove the line I've linked above.

**Review by Kilo59 — CHANGES_REQUESTED** [review]
Type-Checking

https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L146

File: great_expectations/validator/validation_graph.py
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py:1]: @Kilo59 I have done it and committed -- now trying to reopen this PR, which I accidentally closed.  Thanks.

**Comment by alexsherstinsky** [comment]
Trying to reopen this pull request.

**Comment by alexsherstinsky** [comment]
The content of this (accidentally closed) pull request has been moved to #6228 which is presently open.  Thank you.

**Pull request closed without merging**
